File: apps/readest-app/src-tauri/plugins/tauri-plugin-native-bridge/ios/Sources/ClipUrlController.swift
import UIKit
import WebKit
import os

private let clipLogger = Logger(subsystem: Bundle.main.bundleIdentifier!, category: "ClipUrl")

/// Args decoded from the JS `invoke('clip_url', { ... })` payload.
/// Mirrors `ClipOptions` in `clip_url.rs` field-for-field (camelCase
/// JSON, all-but-`url` optional). The defaults below match the Rust
/// `ClipOptions::*()` accessors — English / dark-palette — so a
/// caller that omits a field still renders sensibly.
class ClipUrlArgs: Decodable {
  let url: String
  let windowTitle: String?
  let overlayTitle: String?
  let loadingStatus: String?
  let capturingStatus: String?
  let savedTitle: String?
  let background: String?
  let foreground: String?
  // Interactive mode: show the page with a Cancel/Capture bar instead of
  // the opaque overlay so the user can sign in before capturing.
  let interactive: Bool?
  let signInHint: String?
  let captureLabel: String?
  let cancelLabel: String?

  // Resolved getters with the same fallbacks the Rust impl uses.
  var resolvedOverlayTitle: String { overlayTitle ?? "Saving to Readest" }
  var resolvedLoadingStatus: String { loadingStatus ?? "Loading article…" }
  var resolvedCapturingStatus: String { capturingStatus ?? "Capturing article…" }
  var resolvedBackground: String { background ?? "#1f2024" }
  var resolvedForeground: String { foreground ?? "#f5f5f7" }
  var resolvedSignInHint: String { signInHint ?? "Sign in if needed, then capture" }
  var resolvedCaptureLabel: String { captureLabel ?? "Capture" }
  var resolvedCancelLabel: String { cancelLabel ?? "Cancel" }
}

/// Errors surfaced to the JS caller through `invoke.reject`. Strings
/// match the desktop `clip_url` error vocabulary so callers handling
/// the rejection don't need a separate mobile branch.
enum ClipUrlError: Error {
  case invalidUrl
  case loadFailed(String)
  case timedOut
  // User closed the interactive sign-in capture. Shared cancel vocabulary
  // with `ClipUrlController.kt` — the JS side matches this string to stay
  // quiet on user-initiated cancels.
  case cancelled

  var message: String {
    switch self {
    case .invalidUrl: return "Invalid URL"
    case .loadFailed(let detail): return "Could not fetch this page: \(detail)"
    case .timedOut: return "Page took too long to load"
    case .cancelled: return "Capture cancelled"
    }
  }
}

/// Full-screen view controller that loads `options.url` in a WKWebView,
/// shows a deliberate "Saving…" overlay over the article render so the
/// user sees expected progress (not a website flashing by), and once
/// `webView(_:didFinish:)` fires + a 3 s settle window passes — long
/// enough for in-flight lazy-loaders / Cloudflare-style JS challenges
/// to resolve — captures `document.documentElement.outerHTML` and
/// hands it back via the completion handler.
///
/// Mirrors the desktop `clip_url` flow shape: same Chrome UA, same
/// fingerprint mask, same overlay (rendered via Swift here instead of
/// an inline JS user-script — Swift gives us a proper UIView spinner
/// that doesn't get wiped by the page's own hydration).
final class ClipUrlController: UIViewController, WKNavigationDelegate {

  // Real Chrome UA. Tauri's default reports Safari/WKWebView, which
  // sites with aggressive bot detection cross-check against
  // `navigator.*` fingerprints and reject. Same string as the desktop
  // `BROWSER_UA` constant in `clip_url.rs`.
  static let browserUserAgent =
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 "
    + "(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"

  // Total budget from load-start to outerHTML capture. Same as desktop.
  static let hardTimeoutSeconds: TimeInterval = 30
  // Settle delay after `didFinish` so IntersectionObserver-based lazy
  // loaders fire for content already in the viewport.
  static let loadSettleSeconds: TimeInterval = 3

  private let args: ClipUrlArgs
  private let completion: (Result<String, ClipUrlError>) -> Void

  private var webView: WKWebView!
  private var overlayView: UIView!
  private var statusLabel: UILabel!
  private var didFinishOrFail = false
  private var captureFired = false
  private var timeoutWorkItem: DispatchWorkItem?

  init(args: ClipUrlArgs, completion: @escaping (Result<String, ClipUrlError>) -> Void) {
    self.args = args
    self.completion = completion
    super.init(nibName: nil, bundle: nil)
    // Modal full-screen so the overlay covers the chrome and the app
    // behind it doesn't peek through during the brief capture window.
    self.modalPresentationStyle = .fullScreen
  }

  required init?(coder: NSCoder) {
    fatalError("init(coder:) is not supported for ClipUrlController")
  }

  private var interactiveMode: Bool { args.interactive == true }

  override func viewDidLoad() {
    super.viewDidLoad()
    view.backgroundColor = UIColor(hexString: args.resolvedBackground) ?? .black
    if interactiveMode {
      // Interactive: the page is visible under a Cancel/Capture bar so
      // the user can sign in first. The WKWebsiteDataStore is the app's
      // default (persistent) store, so the session survives for every
      // later headless clip.
      let bar = setUpActionBar()
      setUpWebView(below: bar)
    } else {
      setUpWebView(below: nil)
      setUpOverlay()
    }
    startCapture()
  }

  // MARK: - Setup

  private func setUpWebView(below bar: UIView?) {
    let config = WKWebViewConfiguration()
    // Inject the same fingerprint-mask script the desktop flow uses,
    // before any page script runs, so `navigator.webdriver` and the
    // window.chrome shape look real to first-party detection code.
    let mask = WKUserScript(
      source: ClipUrlController.fingerprintMaskScript(),
      injectionTime: .atDocumentStart,
      forMainFrameOnly: false
    )
    config.userContentController.addUserScript(mask)

    config.allowsInlineMediaPlayback = true
    config.mediaTypesRequiringUserActionForPlayback = .all
    // Suppress text selection / link previews — purely cosmetic; the
    // overlay covers the page anyway. Defensive, in case the spinner
    // catches a stray tap.
    config.preferences.javaScriptCanOpenWindowsAutomatically = false

    let wv = WKWebView(frame: .zero, configuration: config)
    wv.customUserAgent = ClipUrlController.browserUserAgent
    wv.navigationDelegate = self
    wv.allowsBackForwardNavigationGestures = false
    wv.isOpaque = true
    wv.backgroundColor = UIColor(hexString: args.resolvedBackground) ?? .black
    wv.scrollView.backgroundColor = wv.backgroundColor
    wv.translatesAutoresizingMaskIntoConstraints = false
    view.addSubview(wv)
    NSLayoutConstraint.activate([
      wv.topAnchor.constraint(equalTo: bar?.bottomAnchor ?? view.topAnchor),
      wv.leadingAnchor.constraint(equalTo: view.leadingAnchor),
      wv.trailingAnchor.constraint(equalTo: view.trailingAnchor),
      wv.bottomAnchor.constraint(equalTo: view.bottomAnchor),
    ])
    self.webView = wv
  }

  /// Interactive-mode top bar: hint text + Cancel + a solid Capture
  /// button, drawn with the caller's theme colours and extending under
  /// the status bar (content laid out inside the safe area).
  private func setUpActionBar() -> UIView {
    let bg = UIColor(hexString: args.resolvedBackground) ?? .black
    let fg = UIColor(hexString: args.resolvedForeground) ?? .white

    let bar = UIView()
    bar.backgroundColor = bg
    bar.translatesAutoresizingMaskIntoConstraints = false
    view.addSubview(bar)

    let hint = UILabel()
    hint.text = args.resolvedSignInHint
    hint.textColor = fg.withAlphaComponent(0.7)
    hint.font = UIFont.systemFont(ofSize: 13)
    hint.numberOfLines = 2
    hint.lineBreakMode = .byTruncatingTail
    hint.setContentCompressionResistancePriority(.defaultLow, for: .horizontal)
    hint.translatesAutoresizingMaskIntoConstraints = false
    bar.addSubview(hint)

    let cancel = UIButton(type: .system)
    cancel.setTitle(args.resolvedCancelLabel, for: .normal)
    cancel.setTitleColor(fg, for: .normal)
    cancel.titleLabel?.font = UIFont.systemFont(ofSize: 15)
    cancel.addTarget(self, action: #selector(cancelTapped), for: .touchUpInside)
    cancel.translatesAutoresizingMaskIntoConstraints = false
    bar.addSubview(cancel)

    let capture = UIButton(type: .system)
    capture.setTitle(args.resolvedCaptureLabel, for: .normal)
    capture.setTitleColor(bg, for: .normal)
    capture.titleLabel?.font = UIFont.systemFont(ofSize: 15, weight: .semibold)
    capture.backgroundColor = fg
    capture.layer.cornerRadius = 16
    capture.contentEdgeInsets = UIEdgeInsets(top: 6, left: 16, bottom: 6, right: 16)
    capture.addTarget(self, action: #selector(captureTapped), for: .touchUpInside)
    capture.translatesAutoresizingMaskIntoConstraints = false
    bar.addSubview(capture)

    NSLayoutConstraint.activate([
      bar.topAnchor.constraint(equalTo: view.topAnchor),
      bar.leadingAnchor.constraint(equalTo: view.leadingAnchor),
      bar.trailingAnchor.constraint(equalTo: view.trailingAnchor),

      hint.leadingAnchor.constraint(equalTo: bar.leadingAnchor, constant: 16),
      hint.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 8),
      hint.bottomAnchor.constraint(equalTo: bar.bottomAnchor, constant: -8),
      hint.centerYAnchor.constraint(equalTo: capture.centerYAnchor),

      cancel.leadingAnchor.constraint(greaterThanOrEqualTo: hint.trailingAnchor, constant: 12),
      cancel.centerYAnchor.constraint(equalTo: capture.centerYAnchor),

      capture.leadingAnchor.constraint(equalTo: cancel.trailingAnchor, constant: 12),
      capture.trailingAnchor.constraint(equalTo: bar.trailingAnchor, constant: -12),
    ])
    return bar
  }

  @objc private func cancelTapped() {
    finish(.failure(.cancelled))
  }

  @objc private func captureTapped() {
    captureOuterHtml()
  }

  private func setUpOverlay() {
    let bg = UIColor(hexString: args.resolvedBackground) ?? .black
    let fg = UIColor(hexString: args.resolvedForeground) ?? .white
    view.backgroundColor = bg

    let overlay = UIView()
    overlay.backgroundColor = bg
    overlay.translatesAutoresizingMaskIntoConstraints = false
    view.addSubview(overlay)
    NSLayoutConstraint.activate([
      overlay.topAnchor.constraint(equalTo: view.topAnchor),
      overlay.leadingAnchor.constraint(equalTo: view.leadingAnchor),
      overlay.trailingAnchor.constraint(equalTo: view.trailingAnchor),
      overlay.bottomAnchor.constraint(equalTo: view.bottomAnchor),
    ])

    let spinner = UIActivityIndicatorView(style: .large)
    spinner.color = fg.withAlphaComponent(0.85)
    spinner.translatesAutoresizingMaskIntoConstraints = false
    spinner.startAnimating()
    overlay.addSubview(spinner)

    let title = UILabel()
    title.text = args.resolvedOverlayTitle
    title.textColor = fg
    title.font = UIFont.systemFont(ofSize: 15, weight: .semibold)
    title.textAlignment = .center
    title.translatesAutoresizingMaskIntoConstraints = false
    overlay.addSubview(title)

    let status = UILabel()
    status.text = args.resolvedLoadingStatus
    status.textColor = fg.withAlphaComponent(0.7)
    status.font = UIFont.systemFont(ofSize: 13)
    status.textAlignment = .center
    status.numberOfLines = 1
    status.lineBreakMode = .byTruncatingTail
    status.translatesAutoresizingMaskIntoConstraints = false
    overlay.addSubview(status)
    self.statusLabel = status

    NSLayoutConstraint.activate([
      spinner.centerXAnchor.constraint(equalTo: overlay.centerXAnchor),
      spinner.centerYAnchor.constraint(equalTo: overlay.centerYAnchor, constant: -28),
      title.topAnchor.constraint(equalTo: spinner.bottomAnchor, constant: 14),
      title.centerXAnchor.constraint(equalTo: overlay.centerXAnchor),
      title.leadingAnchor.constraint(greaterThanOrEqualTo: overlay.leadingAnchor, constant: 24),
      title.trailingAnchor.constraint(lessThanOrEqualTo: overlay.trailingAnchor, constant: -24),
      status.topAnchor.constraint(equalTo: title.bottomAnchor, constant: 4),
      status.centerXAnchor.constraint(equalTo: overlay.centerXAnchor),
      status.leadingAnchor.constraint(greaterThanOrEqualTo: overlay.leadingAnchor, constant: 24),
      status.trailingAnchor.constraint(lessThanOrEqualTo: overlay.trailingAnchor, constant: -24),
    ])
    self.overlayView = overlay
  }

  // MARK: - Capture flow

  private func startCapture() {
    guard let url = URL(string: args.url),
      let scheme = url.scheme?.lowercased(),
      scheme == "http" || scheme == "https"
    else {
      finish(.failure(.invalidUrl))
      return
    }

    var req = URLRequest(url: url)
    req.setValue(ClipUrlController.browserUserAgent, forHTTPHeaderField: "User-Agent")
    webView.load(req)

    // Interactive capture waits for the user's tap — no deadline.
    if interactiveMode { return }

    // Hard timeout — fires even if `didFinish` never does (SPA, redirect
    // chain, JS challenge that never resolves). Same 30 s budget as the
    // desktop flow.
    let work = DispatchWorkItem { [weak self] in
      guard let self = self, !self.captureFired else { return }
      clipLogger.warning("clip_url: hard timeout after \(ClipUrlController.hardTimeoutSeconds, privacy: .public)s")
      self.finish(.failure(.timedOut))
    }
    timeoutWorkItem = work
    DispatchQueue.main.asyncAfter(
      deadline: .now() + ClipUrlController.hardTimeoutSeconds, execute: work)
  }

  func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
    // Interactive capture is user-driven: the page keeps navigating
    // (login redirects) until Capture is tapped.
    guard !interactiveMode else { return }
    guard !didFinishOrFail else { return }
    didFinishOrFail = true
    statusLabel.text = args.resolvedCapturingStatus
    // Settle then capture. Matches the 3 s post-load delay in the
    // desktop init script — long enough for in-flight asset fetches
    // and lazy-load IntersectionObservers to fire.
    DispatchQueue.main.asyncAfter(deadline: .now() + ClipUrlController.loadSettleSeconds) {
      [weak self] in
      self?.captureOuterHtml()
    }
  }

  func webView(
    _ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error
  ) {
    // A failed hop mid-sign-in shouldn't tear the flow down — the page
    // renders its own error and the user can retry or cancel.
    if interactiveMode { return }
    if didFinishOrFail { return }
    didFinishOrFail = true
    finish(.failure(.loadFailed(error.localizedDescription)))
  }

  func webView(
    _ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!,
    withError error: Error
  ) {
    if interactiveMode { return }
    if didFinishOrFail { return }
    didFinishOrFail = true
    finish(.failure(.loadFailed(error.localizedDescription)))
  }

  private func captureOuterHtml() {
    guard !captureFired else { return }
    captureFired = true
    webView.evaluateJavaScript("document.documentElement.outerHTML") { [weak self] result, error in
      guard let self = self else { return }
      if let html = result as? String, !html.isEmpty {
        clipLogger.log("clip_url: captured \(html.count, privacy: .public) chars")
        self.finish(.success(html))
      } else {
        let detail = error?.localizedDescription ?? "empty HTML"
        self.finish(.failure(.loadFailed(detail)))
      }
    }
  }

  private func finish(_ result: Result<String, ClipUrlError>) {
    timeoutWorkItem?.cancel()
    timeoutWorkItem = nil
    // Stop the WebView before dismissing — otherwise its JS keeps
    // running in the background until ARC tears it down, which on
    // older devices can stutter the dismiss animation.
    webView?.stopLoading()
    webView?.navigationDelegate = nil

    dismiss(animated: true) { [completion] in
      completion(result)
    }
  }

  // MARK: - Inline JS scripts

  /// Same shape as `fingerprint_mask_script()` in `clip_url.rs`. Clears
  /// the `navigator.webdriver` / window.chrome / navigator.languages
  /// signals that obvious bot-detection scripts probe.
  static func fingerprintMaskScript() -> String {
    return """
      (function() {
        try {
          Object.defineProperty(navigator, 'webdriver', { get: () => undefined });
        } catch (e) {}
        try {
          if (!window.chrome) { window.chrome = { runtime: {} }; }
        } catch (e) {}
        try {
          if (navigator.languages && navigator.languages.length === 0) {
            Object.defineProperty(navigator, 'languages', { get: () => ['en-US', 'en'] });
          }
        } catch (e) {}
      })();
      """
  }
}

// MARK: - UIColor hex helper

private extension UIColor {
  /// Parse `#rrggbb` (optionally without the `#`) into a UIColor.
  /// Mirrors the Rust-side `parse_hex_color` semantics: returns nil for
  /// anything malformed so the caller falls back to its own default.
  convenience init?(hexString: String) {
    var hex = hexString.trimmingCharacters(in: .whitespacesAndNewlines)
    if hex.hasPrefix("#") { hex = String(hex.dropFirst()) }
    guard hex.count == 6, let v = UInt32(hex, radix: 16) else { return nil }
    let r = CGFloat((v >> 16) & 0xff) / 255.0
    let g = CGFloat((v >> 8) & 0xff) / 255.0
    let b = CGFloat(v & 0xff) / 255.0
    self.init(red: r, green: g, blue: b, alpha: 1.0)
  }
}

File: apps/readest-app/src-tauri/plugins/tauri-plugin-native-bridge/android/src/main/java/ClipUrlController.kt
package com.readest.native_bridge

import android.app.Activity
import android.app.Dialog
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.GradientDrawable
import android.os.Handler
import android.os.Looper
import android.text.TextUtils
import android.util.Log
import android.view.Gravity
import android.view.KeyEvent
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.view.WindowManager
import android.webkit.CookieManager
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import app.tauri.annotation.InvokeArg
import org.json.JSONObject
import java.lang.ref.WeakReference

/**
 * Args decoded from the JS `invoke('clip_url', { ... })` payload.
 * Mirrors `ClipOptions` in `clip_url.rs` field-for-field. Defaults
 * are applied in `ClipUrlController` so a caller that omits a field
 * still renders sensibly.
 */
@InvokeArg
class ClipUrlArgs {
    var url: String? = null
    var windowTitle: String? = null
    var overlayTitle: String? = null
    var loadingStatus: String? = null
    var capturingStatus: String? = null
    var savedTitle: String? = null
    var background: String? = null
    var foreground: String? = null
    // Interactive mode: show the page with a Cancel/Capture bar instead
    // of the opaque overlay so the user can sign in before capturing.
    var interactive: Boolean? = null
    var signInHint: String? = null
    var captureLabel: String? = null
    var cancelLabel: String? = null
}

/**
 * Result handed back to the calling [NativeBridgePlugin.clip_url] —
 * either the captured `document.documentElement.outerHTML`, or one of
 * the error strings the JS layer surfaces verbatim. The error
 * vocabulary matches the desktop `clip_url` Rust impl so callers
 * handling the rejection don't need a mobile-specific branch.
 */
sealed class ClipUrlResult {
    data class Success(val html: String) : ClipUrlResult()
    data class Failure(val message: String) : ClipUrlResult()
}

/**
 * Full-screen Dialog that loads `args.url` in a `WebView`, shows a
 * "Saving…" overlay over the article render (so the user sees a
 * deliberate progress state, not a website flashing by), waits for
 * `onPageFinished` + a 3 s settle window, then captures
 * `document.documentElement.outerHTML` via `evaluateJavascript`.
 *
 * Mirrors the desktop `clip_url` flow shape: same Chrome UA, same
 * fingerprint mask, same overlay (rendered as native views here
 * instead of an injected user script — gives us a real spinner that
 * the page's own hydration can't wipe).
 */
class ClipUrlController(
    activity: Activity,
    private val args: ClipUrlArgs,
    private val completion: (ClipUrlResult) -> Unit,
) {
    // Hold the Activity weakly: the controller outlives the synchronous
    // command call (it waits up to 30 s for the page), and a strong ref
    // would leak the Activity (and its WebView) if it is destroyed while
    // a clip is in flight.
    private val activityRef = WeakReference(activity)

    companion object {
        private const val TAG = "ClipUrl"

        // Real Chrome UA — same string as the Rust desktop flow uses.
        const val BROWSER_USER_AGENT =
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 " +
            "(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"

        // Total budget from load-start to outerHTML capture, in ms.
        const val HARD_TIMEOUT_MS: Long = 30_000
        // Settle delay after onPageFinished so IntersectionObserver-based
        // lazy-loaders fire for content already in the viewport.
        const val LOAD_SETTLE_MS: Long = 3_000

        // Mirrors `fingerprint_mask_script()` in `clip_url.rs`. Clears
        // the obvious bot-detection signals before any page script runs.
        private const val FINGERPRINT_MASK_JS = """
            (function() {
              try {
                Object.defineProperty(navigator, 'webdriver', { get: () => undefined });
              } catch (e) {}
              try {
                if (!window.chrome) { window.chrome = { runtime: {} }; }
              } catch (e) {}
              try {
                if (navigator.languages && navigator.languages.length === 0) {
                  Object.defineProperty(navigator, 'languages', { get: () => ['en-US', 'en'] });
                }
              } catch (e) {}
            })();
        """

        // Defaults match `ClipOptions::*()` accessors on the Rust side.
        private const val DEFAULT_OVERLAY_TITLE = "Saving to Readest"
        private const val DEFAULT_LOADING_STATUS = "Loading article…"
        private const val DEFAULT_CAPTURING_STATUS = "Capturing article…"
        private const val DEFAULT_BACKGROUND = "#1f2024"
        private const val DEFAULT_FOREGROUND = "#f5f5f7"
        private const val DEFAULT_SIGN_IN_HINT = "Sign in if needed, then capture"
        private const val DEFAULT_CAPTURE_LABEL = "Capture"
        private const val DEFAULT_CANCEL_LABEL = "Cancel"

        // Shared cancel vocabulary with `ClipUrlController.swift` — the JS
        // side matches this string to stay quiet on user-initiated cancels.
        const val CANCELLED_MESSAGE = "Capture cancelled"
    }

    private val mainHandler = Handler(Looper.getMainLooper())
    private val interactiveMode = args.interactive == true
    private var dialog: Dialog? = null
    private var webView: WebView? = null
    private var statusLabel: TextView? = null
    private var didFinishOrFail = false
    private var captureFired = false
    private var settled = false
    private val timeoutRunnable = Runnable { onTimeout() }

    fun show() {
        val urlStr = args.url
        if (urlStr.isNullOrBlank() || !(urlStr.startsWith("http://") || urlStr.startsWith("https://"))) {
            completion(ClipUrlResult.Failure("Invalid URL"))
            return
        }
        val act = activityRef.get()
        if (act == null || act.isFinishing || act.isDestroyed) {
            completion(ClipUrlResult.Failure("Activity is no longer available"))
            return
        }
        mainHandler.post { presentDialog(act, urlStr) }
    }

    private fun presentDialog(act: Activity, urlStr: String) {
        val bg = parseHexColor(args.background ?: DEFAULT_BACKGROUND) ?: Color.BLACK
        val fg = parseHexColor(args.foreground ?: DEFAULT_FOREGROUND) ?: Color.WHITE

        // Headless: full-screen dialog with no chrome — we draw our own
        // overlay on top so the underlying app doesn't peek through during
        // the brief capture window. Interactive: keep the status bar (the
        // fullscreen theme breaks adjustResize, and the user has to type
        // credentials) and let the user drive the page.
        val theme = if (interactiveMode) {
            android.R.style.Theme_Black_NoTitleBar
        } else {
            android.R.style.Theme_Black_NoTitleBar_Fullscreen
        }
        val dlg = Dialog(act, theme)
        dlg.setCancelable(interactiveMode)
        dlg.setCanceledOnTouchOutside(false)
        dlg.window?.also { window ->
            window.setBackgroundDrawable(ColorDrawable(bg))
        }

        val wv = WebView(act)
        configureWebView(wv)

        val root: View
        if (interactiveMode) {
            dlg.setOnCancelListener { finish(ClipUrlResult.Failure(CANCELLED_MESSAGE)) }
            dlg.window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE)
            // Back navigates the page history first — sign-in flows hop
            // through several redirects — and only cancels once exhausted.
            dlg.setOnKeyListener { _, keyCode, event ->
                if (keyCode == KeyEvent.KEYCODE_BACK &&
                    event.action == KeyEvent.ACTION_UP && wv.canGoBack()
                ) {
                    wv.goBack()
                    true
                } else {
                    false
                }
            }
            val column = LinearLayout(act)
            column.orientation = LinearLayout.VERTICAL
            column.setBackgroundColor(bg)
            column.addView(
                buildActionBar(act, bg, fg),
                LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT,
                ),
            )
            column.addView(
                wv,
                LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1f),
            )
            root = column
        } else {
            val frame = FrameLayout(act)
            frame.setBackgroundColor(bg)
            // Reserve a logical size for layout; the WebView is hidden
            // behind the opaque overlay anyway, but it still needs a
            // non-zero rect for the page to fire its viewport-based
            // lazy-loaders.
            wv.layoutParams = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT,
            )
            frame.addView(wv)
            val overlay = buildOverlay(act, bg, fg)
            overlay.layoutParams = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT,
            )
            frame.addView(overlay)
            root = frame
        }

        dlg.setContentView(
            root,
            ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT,
            ),
        )
        dlg.window?.setLayout(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT,
        )
        dlg.show()

        dialog = dlg
        webView = wv

        // Inject the fingerprint mask before the first navigation so
        // navigator.webdriver and friends look right when the page's
        // own scripts run.
        wv.evaluateJavascript(FINGERPRINT_MASK_JS, null)
        wv.loadUrl(urlStr)

        // Interactive capture waits for the user's tap — no deadline.
        if (!interactiveMode) {
            mainHandler.postDelayed(timeoutRunnable, HARD_TIMEOUT_MS)
        }
    }

    private fun configureWebView(wv: WebView) {
        val settings: WebSettings = wv.settings
        settings.javaScriptEnabled = true
        settings.domStorageEnabled = true
        settings.databaseEnabled = true
        settings.userAgentString = BROWSER_USER_AGENT
        settings.loadsImagesAutomatically = true
        settings.mediaPlaybackRequiresUserGesture = true
        settings.mixedContentMode = WebSettings.MIXED_CONTENT_COMPATIBILITY_MODE
        wv.setBackgroundColor(parseHexColor(args.background ?: DEFAULT_BACKGROUND) ?: Color.BLACK)

        // The CookieManager is app-wide and persistent, so a session the
        // user establishes in the interactive capture flow authenticates
        // every later headless clip. Third-party cookies are required by
        // most SSO redirects (#5262).
        val cookies = CookieManager.getInstance()
        cookies.setAcceptCookie(true)
        cookies.setAcceptThirdPartyCookies(wv, true)

        wv.webViewClient = object : WebViewClient() {
            override fun onPageFinished(view: WebView?, url: String?) {
                // Interactive capture is user-driven: the page keeps
                // navigating (login redirects) until Capture is tapped.
                if (interactiveMode) return
                if (didFinishOrFail) return
                didFinishOrFail = true
                mainHandler.post {
                    statusLabel?.text = args.capturingStatus ?: DEFAULT_CAPTURING_STATUS
                }
                // Settle then capture. Matches the 3 s post-load delay
                // in the desktop init script.
                mainHandler.postDelayed({ captureOuterHtml() }, LOAD_SETTLE_MS)
            }

            override fun onReceivedError(
                view: WebView?,
                request: WebResourceRequest?,
                error: WebResourceError?,
            ) {
                // A failed hop mid-sign-in shouldn't tear the flow down —
                // the page renders its own error and the user can retry
                // or cancel.
                if (interactiveMode) return
                // Only honour main-frame errors. Subresource failures
                // (a single missing image, a blocked tracker) shouldn't
                // abort the capture — they were noise the desktop flow
                // tolerated too.
                if (request?.isForMainFrame != true) return
                if (didFinishOrFail) return
                didFinishOrFail = true
                val detail = error?.description?.toString() ?: "load failed"
                finish(ClipUrlResult.Failure("Could not fetch this page: $detail"))
            }
        }
    }

    /** Interactive-mode top bar: hint text + Cancel + a solid Capture
     *  button, all drawn with the caller's theme colours. */
    private fun buildActionBar(act: Activity, bg: Int, fg: Int): View {
        val bar = LinearLayout(act)
        bar.orientation = LinearLayout.HORIZONTAL
        bar.gravity = Gravity.CENTER_VERTICAL
        bar.setBackgroundColor(bg)
        bar.setPadding(dp(act, 16), dp(act, 10), dp(act, 12), dp(act, 10))

        val hint = TextView(act)
        hint.text = args.signInHint ?: DEFAULT_SIGN_IN_HINT
        hint.setTextColor(
            Color.argb((0.7f * 255).toInt(), Color.red(fg), Color.green(fg), Color.blue(fg)),
        )
        hint.textSize = 13f
        hint.maxLines = 2
        hint.ellipsize = TextUtils.TruncateAt.END
        val hintParams = LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f)
        hintParams.marginEnd = dp(act, 12)
        hint.layoutParams = hintParams
        bar.addView(hint)

        val cancel = TextView(act)
        cancel.text = args.cancelLabel ?: DEFAULT_CANCEL_LABEL
        cancel.setTextColor(fg)
        cancel.textSize = 14f
        cancel.setPadding(dp(act, 12), dp(act, 8), dp(act, 12), dp(act, 8))
        cancel.setOnClickListener { finish(ClipUrlResult.Failure(CANCELLED_MESSAGE)) }
        bar.addView(cancel)

        val capture = TextView(act)
        capture.text = args.captureLabel ?: DEFAULT_CAPTURE_LABEL
        capture.setTextColor(bg)
        capture.textSize = 14f
        capture.setTypeface(capture.typeface, android.graphics.Typeface.BOLD)
        capture.setPadding(dp(act, 16), dp(act, 8), dp(act, 16), dp(act, 8))
        val pill = GradientDrawable()
        pill.setColor(fg)
        pill.cornerRadius = dp(act, 18).toFloat()
        capture.background = pill
        val captureParams = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT,
        )
        captureParams.marginStart = dp(act, 4)
        capture.layoutParams = captureParams
        capture.setOnClickListener { captureOuterHtml() }
        bar.addView(capture)

        return bar
    }

    private fun buildOverlay(act: Activity, bg: Int, fg: Int): View {
        val column = LinearLayout(act)
        column.orientation = LinearLayout.VERTICAL
        column.gravity = Gravity.CENTER
        column.setBackgroundColor(bg)
        val pad = dp(act, 24)
        column.setPadding(pad, pad, pad, pad)

        val spinner = ProgressBar(act)
        // Tint with foreground at ~85% — same idea as the iOS overlay.
        val spinTint = Color.argb(
            (0.85f * 255).toInt(),
            Color.red(fg), Color.green(fg), Color.blue(fg),
        )
        spinner.indeterminateDrawable?.setTint(spinTint)
        val spinSize = dp(act, 36)
        val spinParams = LinearLayout.LayoutParams(spinSize, spinSize)
        spinParams.bottomMargin = dp(act, 14)
        spinner.layoutParams = spinParams
        column.addView(spinner)

        val title = TextView(act)
        title.text = args.overlayTitle ?: DEFAULT_OVERLAY_TITLE
        title.setTextColor(fg)
        title.textSize = 15f
        title.gravity = Gravity.CENTER
        title.setTypeface(title.typeface, android.graphics.Typeface.BOLD)
        column.addView(title)

        val status = TextView(act)
        status.text = args.loadingStatus ?: DEFAULT_LOADING_STATUS
        // 70% alpha for the secondary line — matches the desktop overlay.
        status.setTextColor(
            Color.argb(
                (0.7f * 255).toInt(),
                Color.red(fg), Color.green(fg), Color.blue(fg),
            ),
        )
        status.textSize = 13f
        status.gravity = Gravity.CENTER
        status.maxLines = 1
        status.ellipsize = TextUtils.TruncateAt.END
        val statusParams = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT,
        )
        statusParams.topMargin = dp(act, 4)
        status.layoutParams = statusParams
        column.addView(status)
        statusLabel = status

        return column
    }

    private fun captureOuterHtml() {
        if (captureFired) return
        captureFired = true
        settled = true
        val wv = webView ?: return finish(ClipUrlResult.Failure("WebView vanished before capture"))
        wv.evaluateJavascript("document.documentElement.outerHTML") { result ->
            // `evaluateJavascript` returns the value JSON-encoded, so
            // an HTML string comes back wrapped in quotes with escapes.
            // Parse via JSONObject to recover the raw HTML.
            val html = try {
                JSONObject("{\"v\":$result}").getString("v")
            } catch (e: Exception) {
                Log.w(TAG, "failed to decode captured HTML", e)
                ""
            }
            if (html.isEmpty()) {
                finish(ClipUrlResult.Failure("Could not fetch this page: empty HTML"))
            } else {
                Log.d(TAG, "captured ${html.length} chars")
                finish(ClipUrlResult.Success(html))
            }
        }
    }

    private fun onTimeout() {
        if (captureFired || settled) return
        Log.w(TAG, "clip_url: hard timeout after ${HARD_TIMEOUT_MS}ms")
        finish(ClipUrlResult.Failure("Page took too long to load"))
    }

    private fun finish(result: ClipUrlResult) {
        mainHandler.removeCallbacks(timeoutRunnable)
        try {
            // Persist any session the page established (interactive
            // sign-in) so later headless clips reuse it.
            CookieManager.getInstance().flush()
        } catch (e: Exception) {
            Log.w(TAG, "error flushing cookies after clip_url", e)
        }
        try {
            webView?.stopLoading()
            webView?.webViewClient = WebViewClient()  // detach our delegate
            dialog?.dismiss()
        } catch (e: Exception) {
            Log.w(TAG, "error tearing down clip_url dialog", e)
        }
        dialog = null
        webView = null
        completion(result)
    }

    private fun dp(act: Activity, units: Int): Int =
        (units * act.resources.displayMetrics.density + 0.5f).toInt()

    /** Parse `#rrggbb` into an Android ARGB int; null on malformed input. */
    private fun parseHexColor(s: String): Int? {
        val hex = s.trim().removePrefix("#")
        if (hex.length != 6) return null
        return try {
            val v = hex.toLong(16).toInt()
            Color.rgb((v shr 16) and 0xff, (v shr 8) and 0xff, v and 0xff)
        } catch (_: NumberFormatException) {
            null
        }
    }
}
